experimental/relativetime/src/format.rs

```diff
@@ -0,0 +1,89 @@
+// (online at: https://github.com/unicode-org/icu4x/blob/main/LICENSE ).
+
+use std::fmt::Write;
+
+use fixed_decimal::{FixedDecimal, Sign};
+use writeable::Writeable;
+
+use crate::{
+    options::{Numeric, RelativeTimeFormatterOptions},
+    relativetime::RelativeTimeFormatter,
+};
+
+pub mod parts {
+    use writeable::Part;
+
+    /// The [`Part`] used by [`FormattedRelativeTime`](crate::FormattedRelativeTime) to mark the
+    /// part of the string that is without a placeholder.
+    pub const LITERAL: Part = Part {
+        category: "relativetime",
+        value: "literal",
+    };
+}
+
+/// An intermediate structure returned by [`RelativeTimeFormatter`](crate::RelativeTimeFormatter).
+/// This structure can be consumed via [`Writeable`](Writeable) trait to a string or buffer.
+pub struct FormattedRelativeTime<'a> {
+    pub(crate) formatter: &'a RelativeTimeFormatter,
+    pub(crate) options: &'a RelativeTimeFormatterOptions,
+    pub(crate) value: FixedDecimal,
+}
+
+impl<'a> Writeable for FormattedRelativeTime<'a> {
+    fn write_to_parts<S: writeable::PartsWrite + ?Sized>(&self, sink: &mut S) -> core::fmt::Result {
+        if self.options.numeric == Numeric::Auto {
+            let relatives = &self.formatter.rt.get().relatives;
+            if let Some(i8_value) = self.value.to_string().parse::<i8>().ok() {
+                if let Some(v) = relatives.get(&i8_value) {
+                    sink.with_part(parts::LITERAL, |s| s.write_str(v))?;
+                    return Ok(());
+                }
+            }
+        }
+
+        let plural_rules_mapping = if self.value.sign() == Sign::Negative {
+            &self.formatter.rt.get().past
+        } else {
+            &self.formatter.rt.get().future
+        };
+        let category = self.formatter.plural_rules.category_for(&self.value);
+        let singular_sub_pattern = match category {
+            icu_plurals::PluralCategory::Zero => &plural_rules_mapping.zero,
+            icu_plurals::PluralCategory::One => &plural_rules_mapping.one,
+            icu_plurals::PluralCategory::Two => &plural_rules_mapping.two,
+            icu_plurals::PluralCategory::Few => &plural_rules_mapping.few,
+            icu_plurals::PluralCategory::Many => &plural_rules_mapping.many,
+            icu_plurals::PluralCategory::Other => &None,
+        };
+
+        // Default to using PluralCategory::Other mapping.
+        let singular_sub_pattern = singular_sub_pattern
+            .as_ref()
+            .unwrap_or(&plural_rules_mapping.other);
+
+        // 255 is used to denote a string without placeholder '{0}'.
+        if singular_sub_pattern.index == 255 {
+            sink.with_part(parts::LITERAL, |s| {
+                s.write_str(&singular_sub_pattern.pattern)
+            })?;
+        } else {
+            let (prefix, suffix) = singular_sub_pattern
+                .pattern
+                .split_at(singular_sub_pattern.index as usize);
```

Comment:
Issue: `split_at` is panicky. It would be nice if there were a lint to catch it. Please see https://github.com/unicode-org/icu4x/pull/2854/files for a workaround

Reply:
Is it okay to use `split_at` with explicitly checking `singular_sub_pattern.index <= pattern.len()` ?

Reply:
It is necessary to check the length before calling `split_at`.

Reply:
Ah yes. Using code similar to the above PR.